lib/Support/Path.cpp

```diff
@@ -63,11 +66,16 @@ CanonicalPathCacheImpl::getCanonicalPath(StringRef Path, StringRef WorkingDir) {
       return It->second;
   }
 
+  StringRef CanonPath;
   llvm::SmallString<PATH_MAX> Buffer;
-  if (llvm::sys::fs::real_path(AbsPath.c_str(), Buffer, false)) {
-    return CanonicalFilePathRef::getAsCanonicalPath(AbsPath);
+  if (IsRelative) {
+    CanonPath = AbsPath;
```

Comment:
IIUC this treats all relative paths as "canonical", which seems surprising. I would expect symlinks to be resolved, redundant path components to be trimmed, etc. as with an absolute path.  Even the idea of returning a relative path from something named `getCanonicalPath` is surprising enough that I think it deserves a doc comment at least.  Implementation-wise if you wanted a canonical relative path you could form the absolute path, `real_path` it, and remove the leading working directory if it matches. If the real path is outside the working directory I would expect to get an absolute path.

Reply:
I was expecting updates to `addUnitOutFilePaths` and `registerUnit` (possibly others) rather than changing `getCanonicalPath`. Ie. we need to stop treating the out file as a path and instead just an identifier. I'd still expect paths to be canonicalized properly though.